File: javaee/samples/testcases/src/testcases/ContainerDB.java
/*
 * Copyright 2022 ICON Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package testcases;

import score.Address;
import score.ArrayDB;
import score.Context;
import score.DictDB;
import score.VarDB;
import score.annotation.External;

import java.math.BigInteger;
import java.util.List;
import java.util.Map;

public class ContainerDB {
    private final VarDB<BigInteger> varInt = Context.newVarDB("var_int", BigInteger.class);
    private final VarDB<String> varStr = Context.newVarDB("var_str", String.class);
    private final VarDB<byte[]> varBytes = Context.newVarDB("var_bytes", byte[].class);
    private final VarDB<Boolean> varBool = Context.newVarDB("var_bool", Boolean.class);
    private final VarDB<Address> varAddr = Context.newVarDB("var_addr", Address.class);

    private final DictDB<String, BigInteger> dictInt = Context.newDictDB("dict_int", BigInteger.class);
    private final DictDB<String, String> dictStr = Context.newDictDB("dict_str", String.class);
    private final DictDB<String, byte[]> dictBytes = Context.newDictDB("dict_bytes", byte[].class);
    private final DictDB<String, Boolean> dictBool = Context.newDictDB("dict_bool", Boolean.class);
    private final DictDB<String, Address> dictAddr = Context.newDictDB("dict_addr", Address.class);

    private final ArrayDB<BigInteger> arrInt = Context.newArrayDB("arr_int", BigInteger.class);
    private final ArrayDB<String> arrStr = Context.newArrayDB("arr_str", String.class);
    private final ArrayDB<byte[]> arrBytes = Context.newArrayDB("arr_bytes", byte[].class);
    private final ArrayDB<Boolean> arrBool = Context.newArrayDB("arr_bool", Boolean.class);
    private final ArrayDB<Address> arrAddr = Context.newArrayDB("arr_addr", Address.class);

    @External(readonly=true)
    public Map<String, Object> getVar(String type) {
        switch (type) {
            case "int":
                return Map.of(type, varInt.get());
            case "str":
                return Map.of(type, varStr.get());
            case "bytes":
                return Map.of(type, varBytes.get());
            case "bool":
                return Map.of(type, varBool.get());
            case "addr":
                return Map.of(type, varAddr.get());
        }
        return Map.of();
    }

    @External(readonly=true)
    public Map<String, Object> getDict(String key, String type) {
        try {
            switch (type) {
                case "int":
                    return Map.of(key, dictInt.get(key));
                case "str":
                    return Map.of(key, dictStr.get(key));
                case "bytes":
                    return Map.of(key, dictBytes.get(key));
                case "bool":
                    return Map.of(key, dictBool.get(key));
                case "addr":
                    return Map.of(key, dictAddr.get(key));
            }
        } catch (NullPointerException e) {
            // just catch NPE when there is no corresponding key in the DictDB
        }
        return Map.of();
    }

    @External(readonly=true)
    public List<Object> getArray(String type) {
        switch (type) {
            case "int": {
                int size = arrInt.size();
                BigInteger[] array = new BigInteger[size];
                for (int i = 0; i < size; i++) {
                    array[i] = arrInt.get(i);
                }
                return List.of((Object[]) array);
            }
            case "str": {
                int size = arrStr.size();
                String[] array = new String[size];
                for (int i = 0; i < size; i++) {
                    var item = arrStr.get(i);
                    array[i] = item == null ? "" : item;
                }
                return List.of((Object[]) array);
            }
            case "bytes": {
                int size = arrBytes.size();
                byte[][] array = new byte[size][];
                for (int i = 0; i < size; i++) {
                    array[i] = arrBytes.get(i);
                }
                return List.of((Object[]) array);
            }
            case "bool": {
                int size = arrBool.size();
                Boolean[] array = new Boolean[size];
                for (int i = 0; i < size; i++) {
                    array[i] = arrBool.get(i);
                }
                return List.of((Object[]) array);
            }
            case "addr": {
                int size = arrAddr.size();
                Address[] array = new Address[size];
                for (int i = 0; i < size; i++) {
                    array[i] = arrAddr.get(i);
                }
                return List.of((Object[]) array);
            }
        }
        return List.of();
    }
}

File: testsuite/data/scores/container_db/container_db.py
from iconservice import *


class ContainerDB(IconScoreBase):
    def __init__(self, db: IconScoreDatabase) -> None:
        super().__init__(db)

        # value_type can be int, str, bytes, bool and Address
        self._var_int = VarDB("var_int", db, value_type=int)
        self._var_str = VarDB("var_str", db, value_type=str)
        self._var_bytes = VarDB("var_bytes", db, value_type=bytes)
        self._var_bool = VarDB("var_bool", db, value_type=bool)
        self._var_addr = VarDB("var_addr", db, value_type=Address)

        self._dict_int = DictDB("dict_int", db, value_type=int)
        self._dict_str = DictDB("dict_str", db, value_type=str)
        self._dict_bytes = DictDB("dict_bytes", db, value_type=bytes)
        self._dict_bool = DictDB("dict_bool", db, value_type=bool)
        self._dict_addr = DictDB("dict_addr", db, value_type=Address)

        self._arr_int = ArrayDB("arr_int", db, value_type=int)
        self._arr_str = ArrayDB("arr_str", db, value_type=str)
        self._arr_bytes = ArrayDB("arr_bytes", db, value_type=bytes)
        self._arr_bool = ArrayDB("arr_bool", db, value_type=bool)
        self._arr_addr = ArrayDB("arr_addr", db, value_type=Address)

    def on_install(self) -> None:
        super().on_install()

    def on_update(self) -> None:
        super().on_update()

    @external(readonly=True)
    def getVar(self, type: str) -> dict:
        if type == 'int':
            return {type: self._var_int.get()}
        elif type == 'str':
            return {type: self._var_str.get()}
        elif type == 'bytes':
            return {type: self._var_bytes.get()}
        elif type == 'bool':
            return {type: self._var_bool.get()}
        elif type == 'addr':
            return {type: self._var_addr.get()}
        return {'error': 'InvalidType'}

    @external
    def setVar(self, v_int: int = None, v_str: str = None, v_bytes: bytes = None,
               v_bool: bool = None, v_addr: Address = None):
        if v_int is not None:
            self._var_int.set(v_int)
        elif v_str is not None:
            self._var_str.set(v_str)
        elif v_bytes is not None:
            self._var_bytes.set(v_bytes)
        elif v_bool is not None:
            self._var_bool.set(v_bool)
        elif v_addr is not None:
            self._var_addr.set(v_addr)

    @external(readonly=True)
    def getDict(self, key: str, type: str) -> dict:
        if type == 'int':
            return {key: self._dict_int[key]}
        elif type == 'str':
            return {key: self._dict_str[key]}
        elif type == 'bytes':
            return {key: self._dict_bytes[key]}
        elif type == 'bool':
            return {key: self._dict_bool[key]}
        elif type == 'addr':
            return {key: self._dict_addr[key]}
        return {'error': 'InvalidType'}

    @external
    def setDict(self, key: str, v_int: int = None, v_str: str = None, v_bytes: bytes = None,
                v_bool: bool = None, v_addr: Address = None):
        if v_int is not None:
            self._dict_int[key] = v_int
        elif v_str is not None:
            self._dict_str[key] = v_str
        elif v_bytes is not None:
            self._dict_bytes[key] = v_bytes
        elif v_bool is not None:
            self._dict_bool[key] = v_bool
        elif v_addr is not None:
            self._dict_addr[key] = v_addr

    @external(readonly=True)
    def getArray(self, type: str) -> list:
        if type == 'int':
            return [item for item in self._arr_int]
        elif type == 'str':
            return [item for item in self._arr_str]
        elif type == 'bytes':
            return [item for item in self._arr_bytes]
        elif type == 'bool':
            return [item for item in self._arr_bool]
        elif type == 'addr':
            return [item for item in self._arr_addr]
        return ['InvalidType']

    @external
    def setArray(self, v_int: int = None, v_str: str = None, v_bytes: bytes = None,
                 v_bool: bool = None, v_addr: Address = None):
        if v_int is not None:
            self._arr_int.put(v_int)
        elif v_str is not None:
            self._arr_str.put(v_str)
        elif v_bytes is not None:
            self._arr_bytes.put(v_bytes)
        elif v_bool is not None:
            self._arr_bool.put(v_bool)
        elif v_addr is not None:
            self._arr_addr.put(v_addr)
